src/autoupdater.cpp

```diff
@@ -0,0 +1,281 @@
+                               + result->errorString();
+            }
+            installerFile.remove();
+            bool wasCanceled = false;
+            if (downloadProcessDialog != nullptr) {
+                wasCanceled = downloadProcessDialog->wasCanceled();
+                downloadProcessDialog->close();
+                downloadProcessDialog->deleteLater();
+                downloadProcessDialog = nullptr;
+            }
+            if (!wasCanceled && QMessageBox::critical(
+                    nullptr, tr("Installer download failed"), errorMessage,
+                    QMessageBox::StandardButton::Retry | QMessageBox::StandardButton::Cancel)
+                                == QMessageBox::StandardButton::Retry) {
+                startDownload();
+            }
+        }
+    } else if (isVersionInfoResult) {
+        onReleaseInfoRequestFinished(result);
+    } else {
+        onInstallerDownloadFinished(result);
+    }
+    result->deleteLater();
+}
+
+bool AutoUpdater::parseReleaseTag(int (&version)[3], const QString &releaseTag) {
+    const QRegularExpressionMatch &match = versionRe.match(releaseTag);
+    if (!match.hasMatch()) {
+        return false;
+    }
+    const QString versionParts[] = {"major", "minor", "patch"};
+    for (int i = 0; i < 3; i++) {
+        QString partStr = match.captured(versionParts[i]);
+        int versionPart;
+        if (partStr.isNull()) {
+            if (i != 2) {
+                return false;
+            }
+            versionPart = 0; // Allow missing patch version
+        } else {
+            bool parseOk = true;
+            versionPart = partStr.toInt(&parseOk);
+            if (!parseOk) {
+                return false;
+            }
+        }
+        version[i] = versionPart;
+    }
+    return true;
+}
+
+void AutoUpdater::startDownload() {
+    if (!downloadUrl.isValid()) {
+        return;
+    }
+    if (haveActualInstallerDownloadUrl) {
+        installerFile.unsetError();
+        if (installerFile.isOpen()) {
+            installerFile.reset();
+        } else {
+            while (!installerFile.open(QFile::WriteOnly)) {
+                if (QMessageBox::critical(
+                        nullptr, tr("Installer download failed"),
+                        tr("Failed to save the Birdtray installer:\n")
+                        + installerFile.errorString(),
+                        QMessageBox::StandardButton::Retry | QMessageBox::StandardButton::Cancel)
+                    != QMessageBox::StandardButton::Retry) {
+                    return;
+                }
+            }
+        }
+        if (downloadProcessDialog == nullptr) {
+            downloadProcessDialog = new UpdateDownloadDialog();
+        } else {
+            downloadProcessDialog->reset();
+        }
+        downloadProcessDialog->show();
+        QNetworkReply* reply = networkAccessManager->get(QNetworkRequest(downloadUrl));
+        connect(reply, &QNetworkReply::downloadProgress, this,
```

Comment:
If you were referring to this connect, each time we create a new get request, a new `QNetworkReply` is generated, so we need to connect to the signal of this new object.

Reply:
Yeah, right.